### PR TITLE
Fix for issue #2192 Dynamic relation mapping crash

### PR DESCRIPTION
**Pull request by vilinskiy-playdayteam**

### CODE REVIEW

**Comment by segiddins** [comment]
Could you please look into the test failures?

**Comment by vilinskiy-playdayteam** [comment]
I fixed tests, they are all passed now. But Travis still returns the following error:
xcodebuild: error: Failed to build workspace RestKit with scheme RestKitTests.
	Reason: The run destination iPad 2 is not valid for Testing the scheme 'RestKitTests'

How can it be fixed?

**Comment by segiddins** [comment]
I don't know why travis is having issues with that test...

File: Code/Network/RKManagedObjectRequestOperation.m [hunk 1]
```diff
@@ -377,11 +377,28 @@ - (RKMappingResult *)refetchedMappingResult
...
+                
+                relationshipValue = results;
+            } else {
+                relationshipValue = [object valueForKeyPath:destinationKeyPath];
+            }
+            
             NSArray *mappingInfos = (mappingInfo.relationshipMappingInfo)[destinationKeyPath];
             for (RKMappingInfo *relationshipMappingInfo in mappingInfos) {
                 NSUInteger index = [mappingInfos indexOfObject:relationshipMappingInfo];
-                id mappedObjectAtIndex = ([relationshipValue respondsToSelector:@selector(objectAtIndex:)]) ? [NSSet setWithObject:relationshipValue[index]] : relationshipValue;
```
Comment (segiddins) [inline Code/Network/RKManagedObjectRequestOperation.m]: why this change?

File: Code/Network/RKManagedObjectRequestOperation.m [hunk 2]
```diff
@@ -377,11 +377,28 @@ - (RKMappingResult *)refetchedMappingResult
...
-            id relationshipValue = [object valueForKeyPath:destinationKeyPath];
+            id relationshipValue = nil;
+            // Objects in collection may have different types, so destination keypath may be not applicable to each of them
+            if([object conformsToProtocol:@protocol(NSFastEnumeration)]) {
+                NSMutableSet* results = [NSMutableSet set];
+                for (id item in object) {
+                    @try {
+                        id value = [item valueForKeyPath:destinationKeyPath];
+                        [results addObject:value];
+                    } @catch(NSException*) {
```
Comment (segiddins) [inline Code/Network/RKManagedObjectRequestOperation.m]: we shouldn't be catching any exceptions

File: Code/Network/RKManagedObjectRequestOperation.m (hunk 1 above)
Comment (vilinskiy-playdayteam) [inline Code/Network/RKManagedObjectRequestOperation.m]: This fix provides proper processing of mapped objects' properties. Moreover, there is no need to wrap relationshipValue into NSSet

Comment (segiddins) [inline Code/Network/RKManagedObjectRequestOperation.m]: please explain _how_ it provides proper processing, versus the existing code

File: Code/Network/RKManagedObjectRequestOperation.m (hunk 2 above)
Comment (vilinskiy-playdayteam) [inline Code/Network/RKManagedObjectRequestOperation.m]: If we don't catch this exception, the case described in the issue #2192 will not work. Do you think that it will be better to overwrite completely RKManagedObjectsFromObjectWithMappingInfo and RKManagedObjectsFromMappingResultWithMappingInfo?

From other side, RKGatherManagedObjectsFromObjectWithRelationshipMapping catches exceptions too

File: Code/Network/RKManagedObjectRequestOperation.m (hunk 1 above)
Comment (vilinskiy-playdayteam) [inline Code/Network/RKManagedObjectRequestOperation.m]: In the original version of code, object, if it was not an array or ordered set, it was wrapped into NSSet. Down to callstack, the method **RKManagedObjectsFromObjectWithMappingInfo** took this *object* and extracted all objects for keypaths in *mappingInfo.relationshipMappingInfo* to process them recursively in other **RKManagedObjectsFromObjectWithMappingInfo** call.

But there can be situation (for example, when RKManagedObjectRequestOperation has two objects of the same class, but with properties of different types), like here:

```objc
@interface Chat : NSManagedObject
@property(nonatomic, strong) NSOrderedSet *chatMessages;
@end

@interface ChatMessage : NSManagedObject
@property(nonatomic, strong) ChatMessageParameter *messageParameter;
@end

@interface ChatMessageParameterText : ChatMessageParameter
@property(nonatomic, strong) NSString *text;
@end

@interface ChatMessageParameterPlace : ChatMessageParameter
@property(nonatomic, strong) Place *place;
@end

@interface Place : NSManagedObject
...
@end
```
and property mapping
```objc
RKDynamicMapping *chatMessageParametersMapping = [[RKDynamicMapping alloc] init];
[chatMessageParametersMapping addMatcher:[RKObjectMappingMatcher matcherWithKeyPath:@"type" expectedValue:@"place" objectMapping:placeMapping]];
[chatMessageParametersMapping addMatcher:[RKObjectMappingMatcher matcherWithKeyPath:@"type" expectedValue:@"text" objectMapping:textMapping]];
```
Let Chat has two messages of **place** and **text** type,
During object request operation that maps Chat from server response, RKManagedObjectsFromObjectWithMappingInfo will be called for NSSet of two entities of classes **ChatMessageParameterText** and **ChatMessageParameterPlace** as *object* param and * mappingInfo* that has *placeMapping* as its *relationshipMappingInfo*.
This will lead to crash in 
```objc
[object valueForKeyPath:destinationKeyPath]
```
because **ChatMessageParameterText** has no **place** relationship.

My solution was to separate *object* that is collection of entities from *object* of other type. 
For collection, I iterate through all its items and try safely to receive values from destination key path using try-catch block. So, **RKManagedObjectsFromObjectWithMappingInfo** traverses safely though all mapped relations and collects all mapped entities.
If I leave wrapping to NSSet in **RKManagedObjectsFromMappingResultWithMappingInfo**, I will have nested collection as *object* param in **RKManagedObjectsFromObjectWithMappingInfo** and the described logic will not work

Thanks for your attention and patience

**Comment by segiddins** [comment]
This looks OK to me, but it's a big enough change that I'd like someone else's eyes on it before I merge.

**Comment by vilinskiy-playdayteam** [comment]
Ok, thanks)

**Comment by vilinskiy-playdayteam** [comment]
Pull request reviewers are welcome
